core/modules/replica/ControlThread.h

```diff
@@ -0,0 +1,368 @@
+namespace qserv {
+namespace replica {
+
+/**
+ * Class ControlThreadError represents exceptions throw by the method
+ * of class ControlThread on various error conditions.
+ */
+class ControlThreadError
+    :   public util::Issue {
+public:
+    ControlThreadError(util::Issue::Context const& ctx,
+                       std::string const& message);
+};
+
+/**
+ * Class ControlThreadStopped represents exceptions throw by subclasses
+ * (or method invoked by subclasses) when running subclass-specific
+ * activities as a response to activity cancellation requests. Note, that
+ * this kind of exception is not considered an error.
+ */
+class ControlThreadStopped
+    :   public std::runtime_error {
+public:
+    ControlThreadStopped()
+        :   std::runtime_error("thread stopped") {
+    }
+};
+
+/**
+ * Class ControlThread is the base class for the Controller-side activities run within
+ * dedicated threads. The class provides a public interface for starting and stopping
+ * the activities, notifying clients on abnormal termination of the activities,
+ * as well as an infrastructure supporting an implementation of these activities.
+ */
+class ControlThread
+    :   public std::enable_shared_from_this<ControlThread> {
+
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<ControlThread> Ptr;
+
+    /// The function type for notifications on the abnormal termination of the threads
+    typedef std::function<void(Ptr)> CallbackType;
+
+    /// The function type for functions used in evaluating user-defined early-termination
+    /// conditions for aborting thread completion tracking
+    typedef std::function<bool(Ptr)> WaitEvaluatorType;
+
+    // Copy semantics is prohibited
+
+    ControlThread(ControlThread const&) = delete;
+    ControlThread& operator=(ControlThread const&) = delete;
+
+    virtual ~ControlThread() = default;
+
+    /// @return a reference to a provider of services
+    ServiceProvider::Ptr const& serviceProvider() const { return _controller->serviceProvider(); }
+
+    /// @return a reference to the Controller
+    Controller::Ptr const& controller() const { return _controller; }
+
+    /// @return the name of the thread
+    std::string name() const { return _name; }
+
+    // @return 'true' if the thread is running
+    bool isRunning() const { return _isRunning.load(); }
+
+    /**
+     * Start a subclass-supplied sequence of actions (virtual method 'run')
+     * within a new thread if it's not running. Note that the lifetime of
+     * the current object is guaranteed to be extended for the duration of
+     * the thread.
+     *
+     * @return
+     *   'true' if the thread was already running at a time when this
+     *   method was called
+     */
+    bool start();
+
+    /**
+     * Stop the thread if it's still running
+     *
+     * @return
+     *   'true' if the thread was already stopped at a time when this
+     *   method was called
+     */
+    bool stop();
```

Comment:
this is effectively `detach`, why not use that name? It would be more clear to those familiar with `std::thread`.

Reply:
Well, class `ControlThread` is not just an instance of `std::thread`. This class uses the C++ threads to execute relevant operations. If you look at the implementation of this class then you'll see that method `ControlThread::start()` lunches `std::thread` (and makes it detached), and method `ControlThread::stop()` is attempting to terminate (gracefully) an instance of `std::thread`. So, tis choice of methods makes a sense.